testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketSslClientRenegotiateTest.java

```diff
@@ -0,0 +1,219 @@
+                                sc, cc});
+                    }
+            }
+        }
+
+        return params;
+    }
+
+    private final SslContext serverCtx;
+    private final SslContext clientCtx;
+
+    private final AtomicReference<Throwable> clientException = new AtomicReference<Throwable>();
+    private final AtomicReference<Throwable> serverException = new AtomicReference<Throwable>();
+
+    private volatile Channel clientChannel;
+    private volatile Channel serverChannel;
+
+    private volatile SslHandler clientSslHandler;
+    private volatile SslHandler serverSslHandler;
+
+    private final TestHandler clientHandler = new TestHandler(clientException);
+
+    private final TestHandler serverHandler = new TestHandler(serverException);
+
+    public SocketSslClientRenegotiateTest(
+            SslContext serverCtx, SslContext clientCtx) {
+        this.serverCtx = serverCtx;
+        this.clientCtx = clientCtx;
+    }
+
+    @Test(timeout = 30000)
+    public void testSslRenegotiationRejected() throws Throwable {
+        run();
+    }
+
+    public void testSslRenegotiationRejected(ServerBootstrap sb, Bootstrap cb) throws Throwable {
+        reset();
+
+        sb.childHandler(new ChannelInitializer<Channel>() {
+            @Override
+            @SuppressWarnings("deprecation")
+            public void initChannel(Channel sch) throws Exception {
+                serverChannel = sch;
+                serverSslHandler = serverCtx.newHandler(sch.alloc());
+
+                sch.pipeline().addLast("ssl", serverSslHandler);
+                sch.pipeline().addLast("handler", serverHandler);
+            }
+        });
+
+        cb.handler(new ChannelInitializer<Channel>() {
+            @Override
+            @SuppressWarnings("deprecation")
+            public void initChannel(Channel sch) throws Exception {
+                clientChannel = sch;
+                clientSslHandler = clientCtx.newHandler(sch.alloc());
+
+                sch.pipeline().addLast("ssl", clientSslHandler);
+                sch.pipeline().addLast("handler", clientHandler);
+            }
+        });
+
+        Channel sc = sb.bind().sync().channel();
+        cb.connect().sync();
+
+        Future<Channel> clientHandshakeFuture = clientSslHandler.handshakeFuture();
+        clientHandshakeFuture.sync();
+
+        String renegotiation = "SSL_RSA_WITH_RC4_128_SHA";
+        clientSslHandler.engine().setEnabledCipherSuites(new String[] { renegotiation });
+        clientSslHandler.renegotiate().await();
+        serverChannel.close().awaitUninterruptibly();
+        clientChannel.close().awaitUninterruptibly();
+        sc.close().awaitUninterruptibly();
+
+        try {
+            if (serverException.get() != null) {
+                throw serverException.get();
+            }
+            fail();
+        } catch (DecoderException e) {
+            assertTrue(e.getCause() instanceof SSLHandshakeException);
```

Comment:
Could you also check if `userEventTriggered()` is invoked with proper exception?